File: entranceAnalyser/frontend/src/keptBboxes/placeTypes.ts
/*
 * Copyright Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

/**
 * Quebec POI place-type taxonomy. Keep in sync with
 * `backend/src/place_types.rs` (`PLACE_TYPES` and `TAG_FALLBACK_SQL`).
 */

/** Reviewer-selectable place types, in dropdown order. */
export const PLACE_TYPES = [
    'university_small',
    'university',
    'cegep',
    'hospital',
    'national_park',
    'shopping_center',
    'sport_stadium',
    'conference_center',
    'airport',
    'train_bus_station',
    'ski_resort',
    'primary_school',
    'secondary_school',
    'regional_park',
    'municipal_park',
    'large_clinic',
    'hospice',
    'elderly_residence',
    'museum_cultural',
    'concert_hall',
    'sport_pitch',
    'sport_center',
    'governmental_office',
    'industrial',
    'other',
] as const;

export type PlaceType = (typeof PLACE_TYPES)[number];

/** Human labels for dropdown and stats tables, keyed by slug. */
export const PLACE_TYPE_LABELS: Record<PlaceType, string> = {
    university_small: 'University (small campus)',
    university: 'University',
    cegep: 'College / CEGEP',
    hospital: 'Hospital',
    national_park: 'National park',
    shopping_center: 'Shopping center',
    sport_stadium: 'Sport stadium',
    conference_center: 'Conference center',
    airport: 'Airport',
    train_bus_station: 'Train or bus station',
    ski_resort: 'Ski resort',
    primary_school: 'Primary school',
    secondary_school: 'Secondary school',
    regional_park: 'Regional park',
    municipal_park: 'Municipal park',
    large_clinic: 'Large clinic',
    hospice: 'Hospice',
    elderly_residence: 'Elderly residence',
    museum_cultural: 'Museum or cultural center',
    concert_hall: 'Concert hall',
    sport_pitch: 'Sport pitch',
    sport_center: 'Sport center',
    governmental_office: 'Governmental office',
    industrial: 'Industrial',
    other: 'Other',
};

/** Labels for the Quebec stats table, in display order (includes `other`). */
export const QUEBEC_PLACE_TYPE_LABELS: ReadonlyArray<readonly [PlaceType | 'other', string]> =
    PLACE_TYPES.map((t) => [t, PLACE_TYPE_LABELS[t]] as const);

function isUniversity(tags: Record<string, string>): boolean {
    return (
        tags['amenity'] === 'university' ||
        tags['education'] === 'university' ||
        tags['building'] === 'university'
    );
}

/**
 * Classify a POI from its OSM tags — same rules (and order) as the SQL
 * fallback in `quebec_place_type_stats`. Returns `null` when no rule
 * matches so the reviewer can pick manually (including `other`).
 * @param tags Raw OSM tags of the picked POI.
 */
export function detectPlaceType(tags: Record<string, string>): PlaceType | null {
    if (tags['branch'] && isUniversity(tags)) return 'university_small';
    if (isUniversity(tags)) return 'university';
    if (tags['amenity'] === 'college' || tags['education'] === 'college') return 'cegep';
    if (tags['amenity'] === 'hospital' || tags['healthcare'] === 'hospital') return 'hospital';
    if (
        tags['boundary'] === 'protected_area' &&
        (tags['protect_class'] === '2' ||
            /parc national|national park|parc marin/i.test(tags['name'] ?? ''))
    ) {
        return 'national_park';
    }
    if (tags['shop'] === 'mall' || tags['landuse'] === 'retail') return 'shopping_center';
    if (tags['leisure'] === 'stadium' || tags['building'] === 'stadium') return 'sport_stadium';
    if (tags['amenity'] === 'conference_centre' || tags['amenity'] === 'events_venue') {
        return 'conference_center';
    }
    if (['aerodrome', 'terminal', 'international'].includes(tags['aeroway'] ?? '')) {
        return 'airport';
    }
    if (
        tags['railway'] === 'station' ||
        tags['railway'] === 'halt' ||
        tags['amenity'] === 'bus_station' ||
        tags['public_transport'] === 'station'
    ) {
        return 'train_bus_station';
    }
    if (tags['landuse'] === 'winter_sports' || tags['piste:type']) return 'ski_resort';
    if (
        tags['amenity'] === 'kindergarten' ||
        (tags['amenity'] === 'school' && ['0', '1'].includes(tags['isced:level'] ?? ''))
    ) {
        return 'primary_school';
    }
    if (tags['amenity'] === 'school' && ['2', '3'].includes(tags['isced:level'] ?? '')) {
        return 'secondary_school';
    }
    if (tags['boundary'] === 'protected_area' && tags['protect_class'] === '3') {
        return 'regional_park';
    }
    if (tags['leisure'] === 'park') return 'municipal_park';
    if (tags['amenity'] === 'clinic' || ['centre', 'clinic'].includes(tags['healthcare'] ?? '')) {
        return 'large_clinic';
    }
    if (tags['healthcare'] === 'hospice' || tags['social_facility'] === 'hospice') {
        return 'hospice';
    }
    if (['nursing_home', 'assisted_living', 'senior'].includes(tags['social_facility'] ?? '')) {
        return 'elderly_residence';
    }
    if (
        tags['tourism'] === 'museum' ||
        tags['amenity'] === 'arts_centre' ||
        tags['building'] === 'museum'
    ) {
        return 'museum_cultural';
    }
    if (
        ['theatre', 'music_venue', 'concert_hall'].includes(tags['amenity'] ?? '') ||
        tags['building'] === 'theatre'
    ) {
        return 'concert_hall';
    }
    if (tags['leisure'] === 'pitch' || tags['leisure'] === 'track') return 'sport_pitch';
    if (
        tags['leisure'] === 'sports_centre' ||
        tags['building'] === 'sports_centre' ||
        tags['building'] === 'sports_hall'
    ) {
        return 'sport_center';
    }
    if (
        tags['office'] === 'government' ||
        tags['building'] === 'government' ||
        tags['amenity'] === 'townhall'
    ) {
        return 'governmental_office';
    }
    if (tags['building'] === 'industrial' || tags['man_made'] === 'works') return 'industrial';
    return null;
}

File: entranceAnalyser/backend/src/place_types.rs
/*
 * Copyright Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

//! Quebec POI place-type taxonomy for reviewer picks and stats.
//!
//! [`PLACE_TYPES`] is the allow-list for `PATCH /poi_pick { place_type }`.
//! [`TAG_FALLBACK_SQL`] mirrors the TypeScript rules in
//! `frontend/src/keptBboxes/placeTypes.ts` (first match wins).

/// Reviewer-selectable place types, in dropdown order.
pub const PLACE_TYPES: [&str; 25] = [
    "university_small",
    "university",
    "cegep",
    "hospital",
    "national_park",
    "shopping_center",
    "sport_stadium",
    "conference_center",
    "airport",
    "train_bus_station",
    "ski_resort",
    "primary_school",
    "secondary_school",
    "regional_park",
    "municipal_park",
    "large_clinic",
    "hospice",
    "elderly_residence",
    "museum_cultural",
    "concert_hall",
    "sport_pitch",
    "sport_center",
    "governmental_office",
    "industrial",
    "other",
];

/// SQL `CASE` body classifying a `tags` jsonb column when the reviewer
/// has not set `place_type`. Keep in sync with `detectPlaceType` in the
/// frontend.
pub const TAG_FALLBACK_SQL: &str = "\
     WHEN tags->>'branch' IS NOT NULL \
       AND (tags->>'amenity' = 'university' \
         OR tags->>'education' = 'university' \
         OR tags->>'building' = 'university') THEN 'university_small' \
     WHEN tags->>'amenity' = 'university' \
       OR tags->>'education' = 'university' \
       OR tags->>'building' = 'university' THEN 'university' \
     WHEN tags->>'amenity' = 'college' \
       OR tags->>'education' = 'college' THEN 'cegep' \
     WHEN tags->>'amenity' = 'hospital' \
       OR tags->>'healthcare' = 'hospital' THEN 'hospital' \
     WHEN tags->>'boundary' = 'protected_area' \
       AND (tags->>'protect_class' = '2' \
         OR tags->>'name' ILIKE '%Parc national%' \
         OR tags->>'name' ILIKE '%National Park%' \
         OR tags->>'name' ILIKE '%parc marin%') THEN 'national_park' \
     WHEN tags->>'shop' = 'mall' \
       OR tags->>'landuse' = 'retail' THEN 'shopping_center' \
     WHEN tags->>'leisure' = 'stadium' \
       OR tags->>'building' = 'stadium' THEN 'sport_stadium' \
     WHEN tags->>'amenity' IN ('conference_centre', 'events_venue') \
       THEN 'conference_center' \
     WHEN tags->>'aeroway' IN ('aerodrome', 'terminal', 'international') \
       THEN 'airport' \
     WHEN tags->>'railway' IN ('station', 'halt') \
       OR tags->>'amenity' = 'bus_station' \
       OR tags->>'public_transport' = 'station' THEN 'train_bus_station' \
     WHEN tags->>'landuse' = 'winter_sports' \
       OR tags ? 'piste:type' THEN 'ski_resort' \
     WHEN tags->>'amenity' = 'kindergarten' \
       OR (tags->>'amenity' = 'school' \
         AND tags->>'isced:level' IN ('0', '1')) THEN 'primary_school' \
     WHEN tags->>'amenity' = 'school' \
       AND tags->>'isced:level' IN ('2', '3') THEN 'secondary_school' \
     WHEN tags->>'boundary' = 'protected_area' \
       AND tags->>'protect_class' = '3' THEN 'regional_park' \
     WHEN tags->>'leisure' = 'park' THEN 'municipal_park' \
     WHEN tags->>'amenity' = 'clinic' \
       OR tags->>'healthcare' IN ('centre', 'clinic') THEN 'large_clinic' \
     WHEN tags->>'healthcare' = 'hospice' \
       OR tags->>'social_facility' = 'hospice' THEN 'hospice' \
     WHEN tags->>'social_facility' IN ('nursing_home', 'assisted_living', 'senior') \
       THEN 'elderly_residence' \
     WHEN tags->>'tourism' = 'museum' \
       OR tags->>'amenity' = 'arts_centre' \
       OR tags->>'building' = 'museum' THEN 'museum_cultural' \
     WHEN tags->>'amenity' IN ('theatre', 'music_venue', 'concert_hall') \
       OR tags->>'building' = 'theatre' THEN 'concert_hall' \
     WHEN tags->>'leisure' IN ('pitch', 'track') THEN 'sport_pitch' \
     WHEN tags->>'leisure' = 'sports_centre' \
       OR tags->>'building' IN ('sports_centre', 'sports_hall') \
       THEN 'sport_center' \
     WHEN tags->>'office' = 'government' \
       OR tags->>'building' = 'government' \
       OR tags->>'amenity' = 'townhall' THEN 'governmental_office' \
     WHEN tags->>'building' = 'industrial' \
       OR tags->>'man_made' = 'works' THEN 'industrial' \
     ELSE 'other'";

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn place_types_are_unique() {
        let mut seen = std::collections::HashSet::new();
        for t in PLACE_TYPES {
            assert!(seen.insert(t), "duplicate place type: {t}");
        }
    }
}
